### PR TITLE
Issue 223: Fix connection leak

### LINKED ISSUE #223: Excessive connections to Segment Store
**Problem description**
I have observed 1000+ TCP connections from a Rust application to the Segment Store of a standalone Pravega instance. At this point, the application no longer accepts new connections or can open new files because the OS does not allow any more open files.

The application that reproduces the problem is an HTTP server that provides HTTP Live Streaming from a Pravega stream. It is a Rust application built on Hyper/Warp. Web browsers using it will initiate requests to read bytes from a Pravega byte stream and then cancel the request. This may be contributing to this problem. It is not clear to me if the application code is leaking the connection or if it is the Pravega client. There is a single instance of the Pravega client factory which is created when the application starts.

https://github.com/pravega/gstreamer-pravega/blob/db74586a627f1f65f26412b979fde18916f32f1b/pravega-video-server/src/main.rs#L220

**Problem location**
Possibly related to connection pooling.

**Suggestions for an improvement**
Unknown.

### PR DESCRIPTION
**Pull request by Tristan1900**
**Change log description**  
This PR adds a shutdown handle to PrefetchRead and prevents it from recycling connections.

The reason is that the background task of prefetching read could hang for a long time if the fetching offset is at the end of stream. If byte stream reader calls `seek`, then PrefetchRead will be destroyed and this background task is left unattended and a connection is leaked. 

However, adding shutdown hook itself is not enough. The connection could have stale message from the server when recycled back to pool. And anyone use that connection from the pool will have to deal with this stale message which is annoying. 

**Purpose of the change**  
Fix #223

**What the code does**  
Adds a shutdown hook and prevents PrefetchRead from recycling connections.

**How to verify it**  
Manually verify that no connection leaked when calling `seek`.

### CODE REVIEW

File: src/segment_reader.rs [hunk 1]
```diff
@@ -450,28 +461,38 @@ impl PrefetchingAsyncSegmentReader {
...
-        let result = reader.read(offset, length).await;
-        sender
-            .send(result)
-            .map_err(|e| warn!("failed to send reply back: {:?}", e))
-            .expect("send reply back");
+        tokio::select! {
+            result = reader.read(offset, length) => {
+                sender
+                .send(result)
+                .map_err(|e| warn!("failed to send reply back: {:?}", e))
```
Comment (claudiofahey) [inline src/segment_reader.rs:489]: This error occurs regularly with pravega-video-server. It should not print the whole event and it should not panic. I had to apply the patch https://github.com/claudiofahey/pravega-client-rust/commit/70f6cf81e9b035fc8fe0ec338753ca001caebe86.

**Review by claudiofahey — APPROVED** [review]
I'm now seeing a new panic in pravega-video-server with this PR:

```
Mar 08 19:34:13.772  INFO request{method=GET path=/scopes/examples/streams/camera1/m3u8 version=HTTP/1.1 remote.addr=127.0.0.1:59278 referer=http://localhost:3030/player?scope=examples&stream=camera1}: pravega_video_server::models: begin_timestamp=1970-01-01T00:00:00.000000000Z (0 ns), end_timestamp=2554-07-21T23:33:56.709551615Z (18446744073709551615 ns)
thread 'tokio-runtime-worker' panicked at 'assertion failed: `(left == right)`
  left: `5388163232`,
 right: `0`: Offset of SegmentRead response different from the request', /home/faheyc/nautilus/gstreamer/gstreamer-pravega/pravega-client-rust/src/segment_reader.rs:297:21
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:495:5
   1: std::panicking::begin_panic_fmt
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:437:5
   2: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   3: <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
   4: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   5: std::panic::catch_unwind
   6: tokio::runtime::task::harness::poll_future
   7: tokio::runtime::task::harness::Harness<T,S>::poll
   8: std::Mar 08 19:34:13.773  INFO pravega_video_server::models: Opened Pravega reader
thread::local::LocalKey<T>::with
   9: tokio::runtime::thread_pool::worker::Context::run_task
  10: tokio::runtime::thread_pool::worker::Context::run
  11: tokio::macros::scoped_tls::ScopedKey<T>::set
  12: tokio::runtime::thread_pool::worker::run
  13: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
  14: std::panicking::try
  15: tokio::runtime::task::harness::Harness<T,S>::poll
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: NoSuchSegment { segment: "examples/camera1-index/0.#epoch.0", operation: "get segment info", error_msg: "unexpected reply" }', src/main.rs:161:82
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:495:5
   1: core::panicking::panic_fmt
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/option.rs:1268:5
   3: <warp::filter::and_then::AndThenFuture<T,F> as core::future::future::Future>::poll
   4: <warp::filter::or::EitherFuture<T,U> as core::future::future::Future>::poll
   5: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
   6: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
   7: <tracing_futures::Instrumented<T> as core::future::future::Future>::poll
   8: hyper::proto::h1::dispatch::Dispatcher<D,Bs,I,T>::poll_catch
   9: <hyper::server::conn::upgrades::UpgradeableConnection<I,S,E> as core::future::future::Future>::poll
  10: <hyper::server::conn::spawn_all::NewSvcTask<I,N,S,E,W> as core::future::future::Future>::poll
  11: tokio::runtime::task::harness::poll_future
  12: tokio::runtime::task::harness::Harness<T,S>::poll
  13: std::thread::local::LocalKey<T>::with
  14: tokio::runtime::thread_pool::worker::Context::run_task
  15: tokio::runtime::thread_pool::worker::Context::run
  16: tokio::macros::scoped_tls::ScopedKey<T>::set
  17: tokio::runtime::thread_pool::worker::run
  18: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
  19: std::panicking::try
  20: tokio::runtime::task::harness::Harness<T,S>::poll
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Let me know if you need a reproduction procedure.

**Review by claudiofahey — CHANGES_REQUESTED** [review]
See my previous comment.

**Comment by claudiofahey** [comment]
Here are the reproduction steps:

```
# Install GStreamer.
sudo apt-get install \
    gstreamer1.0-plugins-base \
    gstreamer1.0-plugins-good \
    gstreamer1.0-plugins-bad \
    gstreamer1.0-plugins-ugly \
    gstreamer1.0-libav \
    libatk1.0-dev \
    libcairo-dev \
    libges-1.0-dev \
    libgstreamer1.0-dev \
    libgstreamer-plugins-base1.0-dev \
    libgstreamer-plugins-bad1.0-dev \
    libgstrtspserver-1.0-dev \
    libgtk2.0-dev \
    libgtk-3-dev \
    libpango1.0-dev

# Clone GStreamer Pravega repo.
git clone --branch excessive_connections2 --recursive https://github.com/pravega/gstreamer-pravega
cd gstreamer-pravega

# Start Pravega. This will store LTS in /tmp, not memory.
cd pravega-docker
export HOST_IP=x.x.x.x
./up.sh
cd ..

# Generate synthetic video data and write to Pravega.
# Let this run for at least 1 minute, then continue to the next step in another window. You can keep this running.
STREAM=video1 scripts/videotestsrc-to-pravega-hls.sh

# Start Pravega Video Server.
scripts/pravega-video-server.sh
```

Open your browser to the following.
http://localhost:3030/player?scope=examples&stream=video1

You will see the video play in the browser.

You will also see the following logged repeatedly in pravega-video-server. It does appear to function correctly otherwise.

```
Mar 11 03:14:52.019  WARN pravega_client::segment_reader: failed to receive reply from background read: channel closed
thread 'tokio-runtime-worker' panicked at 'assertion failed: `(left == right)`
  left: `66508128`,
 right: `66508416`: Offset of SegmentRead response different from the request', /home/faheyc/nautilus/gstreamer/gstreamer-pravega/pravega-client-rust/src/segment_reader.rs:297:21
```

For more details, see https://github.com/pravega/gstreamer-pravega/blob/master/README.md.

**Review by claudiofahey — APPROVED** [review]
I retested this with the reproduction steps and it appears to be working well now. Note that I did not review the code in this PR.

**Review by shrids — APPROVED** [review]
Sorry for the delay @Tristan1900 . I had reviewed the changes last week, no blocking comments from my end.

File: src/raw_client.rs [hunk 2]
```diff
@@ -80,6 +81,16 @@ impl<'a> RawClient<'a> for RawClientImpl<'a> {
             .await
             .context(RequestTimeout {})?;
         let reply = result.context(ReadReply {})?;
+        if reply.get_request_id() != request.get_request_id() {
+            client_connection.connection.invalidate();
+        }
+        ensure!(
+            reply.get_request_id() == request.get_request_id(),
```
Comment (tkaitchuck) [inline src/raw_client.rs:88]: You can just add a return call inside the if above that way you don't have to repeat the condition.

File: src/raw_client.rs [hunk 3]
```diff
@@ -100,6 +111,16 @@ impl<'a> RawClient<'a> for RawClientImpl<'a> {
             .await
             .context(RequestTimeout {})?;
         let reply = result.context(ReadReply {})?;
+        if reply.get_request_id() != request.get_request_id() {
```
Comment (tkaitchuck) [inline src/raw_client.rs:114]: Same

**Review by tkaitchuck — APPROVED** [review]

File: bindings/Cargo.toml [hunk 4]
```diff
@@ -44,7 +44,7 @@ uuid = {version = "0.8", features = ["v4"]}
 futures = "0.3.5"
 derive-new = "0.5"
 #Python bindings
-pyo3 = { features = ["extension-module"], optional = true, version = "0.11.0" }
+pyo3 = { features = ["extension-module"], optional = true, version = "0.13.0" }
```
Comment (shrids) [inline bindings/Cargo.toml:47]: Is pyo3 upgrade related to the connection leak issue ? If not could we have it as a separate PR?

**Pull request merged**